src/target/riscv/riscv.c

```diff
@@ -668,6 +668,16 @@ static int set_trigger(struct target *target, unsigned int idx, riscv_reg_t tdat
 	const uint32_t type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
 	const bool is_mcontrol = type == CSR_TDATA1_TYPE_MCONTROL;
 
+	if (type == CSR_TDATA1_TYPE_MCONTROL) {
+		if (get_field(tdata1_rb, CSR_MCONTROL_TIMING) == CSR_MCONTROL_TIMING_BEFORE)
+			r->halt_before_dpc = true;
+	} else if (type == CSR_TDATA1_TYPE_MCONTROL6) {
+		int hit0 = get_field(tdata1_rb, CSR_MCONTROL6_HIT0);
+		int hit1 = get_field(tdata1_rb, CSR_MCONTROL6_HIT1);
+		if (((hit1 << 1) | hit0) == CSR_MCONTROL6_HIT0_BEFORE)
+			r->halt_before_dpc = true;
```

Comment:
[5.7.12. Match Control Type 6 (`mcontrol6`, at 0x7a1), `hit0` field description]:
> The TM updates this field when the trigger fires. 

Therefore a zero is always read here and `r->halt_before_dpc` is always false.

Please, take a look at `riscv_trigger_detect_hit_bits()`.

Reply:
addressedaddressed

Reply:
addressed